### PR TITLE
sql: support ALTER TABLE SET SCHEMA command

### PR DESCRIPTION
**Pull request by RichardJCai**
sql: support ALTER TABLE SET SCHEMA command

Release note (sql change): Added support for
ALTER TABLE/SEQUENCE/VIEW SET SCHEMA to set the schema of
the table to the target schema.

One must have DROP privilege on the table and CREATE privilege
on the schema to perform the operation.

### CODE REVIEW

File: pkg/sql/alter_table_set_schema.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	n         *tree.AlterTableSetSchema
+}
+
+// AlterTableSetSchema sets the schema for a table, view or sequence.
+// Privileges: DROP on source table/view/sequence, CREATE on destination schema.
+func (p *planner) AlterTableSetSchema(
+	ctx context.Context, n *tree.AlterTableSetSchema,
+) (planNode, error) {
+	tn := n.Name.ToTableName()
+	requiredTableKind := tree.ResolveAnyTableKind
```
Comment (rohany) [inline pkg/sql/alter_table_set_schema.go:38]: Why do you need the kind? Can't you just use tree.ResolveRequireAny (or whatever the default value is)

File: pkg/sql/alter_table_set_schema.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	schemaID := n.tableDesc.GetParentSchemaID()
+
+	// Lookup the the schema we want to set to.
+	exists, res, err := params.p.LogicalSchemaAccessor().GetSchema(
+		ctx, p.txn,
+		p.ExecCfg().Codec,
+		databaseID,
+		n.newSchema,
+	)
+
```
Comment (rohany) [inline pkg/sql/alter_table_set_schema.go:93]: nit usually there isn't a newline before these err checks

File: pkg/sql/alter_table_set_schema.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+	tableDesc.DrainingNames = append(tableDesc.DrainingNames, renameDetails)
+
+	if err := p.writeSchemaChange(
+		ctx, tableDesc, sqlbase.InvalidMutationID, tree.AsStringWithFQNames(n.n, params.Ann()),
+	); err != nil {
+		return err
+	}
+
+	// Set the tableDesc's new schema id to the desired schema's id.
+	n.tableDesc.UnexposedParentSchemaID = desiredSchemaID
```
Comment (rohany) [inline pkg/sql/alter_table_set_schema.go:132]: is there a `SetParentSchemaID` function on the `MutableDescriptor` interface?

File: pkg/sql/alter_table_set_schema.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+	newTbKey := sqlbase.MakeObjectNameKey(ctx, p.ExecCfg().Settings,
+		databaseID, desiredSchemaID, tableDesc.Name).Key(p.ExecCfg().Codec)
+
+	// We update the descriptor to the new name, but also leave the mapping of the
```
Comment (rohany) [inline pkg/sql/alter_table_set_schema.go:137]: This comment feels like a bit of a run on sentence

File: pkg/sql/alter_table_set_schema.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	// old name to the id, so that the name is not reused until the schema changer
+	// has made sure it's not in use any more.
+	b := &kv.Batch{}
+	if params.p.extendedEvalCtx.Tracing.KVTracingEnabled() {
+		log.VEventf(ctx, 2, "CPut %s -> %d", newTbKey, tableDesc.ID)
+	}
+	err = catalogkv.WriteDescToBatch(ctx, p.extendedEvalCtx.Tracing.KVTracingEnabled(),
```
Comment (rohany) [inline pkg/sql/alter_table_set_schema.go:144]: Why are you writing the descriptor here?

File: pkg/sql/alter_table_set_schema.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	if err == nil && exists {
+		return pgerror.Newf(pgcode.DuplicateRelation,
+			"relation %s already exists in schema %s", tableDesc.Name, n.newSchema)
+	} else if err != nil {
+		return err
+	}
+
+	b.CPut(newTbKey, tableDesc.ID, nil)
+
+	err = p.txn.Run(ctx, b)
```
Comment (rohany) [inline pkg/sql/alter_table_set_schema.go:163]: nit can just return this

File: pkg/sql/logictest/testdata/logic_test/set_schema [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+# No error should happen if IF EXISTS is used and the table does not exist.
+statement ok
+ALTER TABLE IF EXISTS does_not_exist SET SCHEMA s2
+
+# Cannot set schema to one that does not exist.
+statement error pq: schema does_not_exist does not exist
+ALTER TABLE t SET SCHEMA does_not_exist
+
+# Ensure we cannot set schema if there is already a table in the schema
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/set_schema:21]: nit no period at the end here

also the comment should be "if there is a table with the same name in the schema"

File: pkg/sql/logictest/testdata/logic_test/set_schema [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+INSERT INTO s1.t VALUES (2)
+
+# Ensure we can't select from the public schema.
+statement error pq: relation "public.t" does not exist
+SELECT * FROM public.t
+
+statement error pq: relation "t" does not exist
+SELECT * FROM t
+
+# Test moving from one user defined schema to another.
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/set_schema:53]: Should also test moving back into the public schema

File: pkg/sql/logictest/testdata/logic_test/set_schema [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+statement ok
+GRANT DROP ON t5 TO testuser
+
+user testuser
+
+statement error pq: user testuser does not have CREATE privilege on schema s1
+ALTER TABLE t5 SET SCHEMA s1
+
+# Set schema should fail if there are dependent views.
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/set_schema:205]: is there a test missing here?

**Review by rohany — CHANGES_REQUESTED** [review]

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52090)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/alter_table_set_schema.go, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-38:-MDVYzRj3v-ky05LEl5d:b-oqn7ep) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L38)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Why do you need the kind? Can't you just use tree.ResolveRequireAny (or whatever the default value is)
</blockquote></details>

The default is ResolveAnyTableKind, but if the syntax provided is ALTER SEQUENCE or ALTER VIEW, we only want to be  able to alter a sequence/view. ALTER TABLE allows us to alter all 3 (I checked with pg)
___
*[pkg/sql/alter_table_set_schema.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-132:-MDV_9UYDVvg_1xNxl-I:b-v5rt8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L132)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

is there a `SetParentSchemaID` function on the `MutableDescriptor` interface?
</blockquote></details>

Nope, but I added one to MutableTableDescriptor
___
*[pkg/sql/alter_table_set_schema.go, line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-137:-MDV_f_654m0pTN7aK42:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L137)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

This comment feels like a bit of a run on sentence
</blockquote></details>

Done.
___
*[pkg/sql/alter_table_set_schema.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-144:-MDV_smm655kaGMLYUJK:b-gv33rd) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L144)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Why are you writing the descriptor here?
</blockquote></details>

I copied the logic here from rename_table.go, should this be somewhere else?
___
*[pkg/sql/logictest/testdata/logic_test/set_schema, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zt-r1-53:-MDVgcQ7Cebx2xy6mPVS:b-8b11j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/logictest/testdata/logic_test/set_schema#L53)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Should also test moving back into the public schema
</blockquote></details>

Added
___
*[pkg/sql/logictest/testdata/logic_test/set_schema, line 205 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zt-r1-205:-MDVg9er-sGMAp7z8imt:b-pz5gio) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/logictest/testdata/logic_test/set_schema#L205)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

is there a test missing here?
</blockquote></details>

Oops I forgot to remove this comment, the test case is above.

**Review by rohany — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52090)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai and @rohany)

___
*[pkg/sql/alter_table_set_schema.go, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-38:-MDVuolg7AqDk0yuxZw1:bw7utq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L38)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

The default is ResolveAnyTableKind, but if the syntax provided is ALTER SEQUENCE or ALTER VIEW, we only want to be  able to alter a sequence/view. ALTER TABLE allows us to alter all 3 (I checked with pg)
</blockquote></details>

Oh i see sg
___
*[pkg/sql/alter_table_set_schema.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-132:-MDVx23S2s0DAMRwbC8_:b-8iokxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L132)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Nope, but I added one to MutableTableDescriptor
</blockquote></details>

It might be worth to move all of this mutation into functions on the MutableTableDescriptor -- the far end goal for this is all is to not have raw manipulation of table descriptors out of the sqlbase package
___
*[pkg/sql/alter_table_set_schema.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-144:-MDVv5kz8sHgwViX2yiv:b5tjhr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L144)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I copied the logic here from rename_table.go, should this be somewhere else?
</blockquote></details>

Hmm, I don't think that this is necessary. calling `writeSchemaChange` does this write already.
___
*[pkg/sql/alter_table_set_schema.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDVvUjqEomyJE7fj1Zq:-MDVvUjqEomyJE7fj1Zr:b-d6nabg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6495ac7c14d537d9e9b803122a559f6e095cd914/pkg/sql/alter_table_set_schema.go#L103)):*
> ```Go
> 	}
> 
> 	desiredSchemaID := res.ID
> ```

You have to look at the resulting kind and make sure that it even has an ID

You also should be validating that the target schema is valid to rename to, i.e it can't be virtual or temporary
___
*[pkg/sql/alter_table_set_schema.go, line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDVvverBtQ-kSL4m8_C:-MDVvverBtQ-kSL4m8_D:b-djjh3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6495ac7c14d537d9e9b803122a559f6e095cd914/pkg/sql/alter_table_set_schema.go#L121)):*
> ```Go
> 	}
> 
> 	renameDetails := sqlbase.NameInfo{
> ```

I wonder if there is a good way to pull some of this out into a function that can do this logic on both `MutableTableDescriptor`s and `MutableTypeDescriptor`s.
___
*[pkg/sql/alter_table_set_schema.go, line 153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDVvmGeAJYDsuylmC9K:-MDVvmGf-Bpjtha3QieB:bpglimf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6495ac7c14d537d9e9b803122a559f6e095cd914/pkg/sql/alter_table_set_schema.go#L153)):*
> ```Go
> 	}
> 
> 	exists, _, err = sqlbase.LookupObjectID(
> ```

It seems wrong to look up that the object exists after you've already written the namespace entry -- why does this work?

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52090)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/alter_table_set_schema.go, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-38:-MDWfz3o67Cc9iVAd5j0:bvm7z8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L38)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Oh i see sg
</blockquote></details>

Resolved
___
*[pkg/sql/alter_table_set_schema.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-132:-MDWFpiFEoKztN0t_38u:bicg4cn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L132)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

It might be worth to move all of this mutation into functions on the MutableTableDescriptor -- the far end goal for this is all is to not have raw manipulation of table descriptors out of the sqlbase package
</blockquote></details>

Can you clarify what you mean by all of this mutation?
___
*[pkg/sql/alter_table_set_schema.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-144:-MDWJpg4De40Z_rAxaJR:b-o7mnla) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L144)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Hmm, I don't think that this is necessary. calling `writeSchemaChange` does this write already.
</blockquote></details>

Yeah you're right, writeSchemaChange does call it, I think we only need one writeSchemaChange call as long as it's in the right place.
___
*[pkg/sql/alter_table_set_schema.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDVvUjqEomyJE7fj1Zq:-MDWeXae6H9EWCCFujkX:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/6495ac7c14d537d9e9b803122a559f6e095cd914/pkg/sql/alter_table_set_schema.go#L103)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

You have to look at the resulting kind and make sure that it even has an ID

You also should be validating that the target schema is valid to rename to, i.e it can't be virtual or temporary
</blockquote></details>

Fixed
___
*[pkg/sql/alter_table_set_schema.go, line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDVvverBtQ-kSL4m8_C:-MDWFzVN8tVyl-oBntF3:b-hoy7qs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6495ac7c14d537d9e9b803122a559f6e095cd914/pkg/sql/alter_table_set_schema.go#L121)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I wonder if there is a good way to pull some of this out into a function that can do this logic on both `MutableTableDescriptor`s and `MutableTypeDescriptor`s.
</blockquote></details>

This is also used in rename_table as well, it can definitely be abstracted - I think I'll try to pull this out while adding the ALTER TYPE SET SCHEMA PR.
___
*[pkg/sql/alter_table_set_schema.go, line 153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDVvmGeAJYDsuylmC9K:-MDWJLIj6ubtBsY1Do8n:b59rjh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6495ac7c14d537d9e9b803122a559f6e095cd914/pkg/sql/alter_table_set_schema.go#L153)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

It seems wrong to look up that the object exists after you've already written the namespace entry -- why does this work?
</blockquote></details>

I think this works because the key doesn't actually get written until the b.CPut which is below.
I'll move it up to make it more clear.
___
*[pkg/sql/logictest/testdata/logic_test/set_schema, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zt-r1-53:-MDWJ7_UE6czsmxdBABf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/logictest/testdata/logic_test/set_schema#L53)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Added
</blockquote></details>

Done.

**Review by rohany — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52090)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai and @rohany)

___
*[pkg/sql/alter_table_set_schema.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-132:-MDXublUCTGQa3uNFgPr:b5jk5lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L132)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Can you clarify what you mean by all of this mutation?
</blockquote></details>

I'm thinking this setting the parent schema, and adding draining names at least
___
*[pkg/sql/alter_table_set_schema.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDVvUjqEomyJE7fj1Zq:-MDXu8pR-F92SyvhXNki:b6k811g) ([raw file](https://github.com/cockroachdb/cockroach/blob/6495ac7c14d537d9e9b803122a559f6e095cd914/pkg/sql/alter_table_set_schema.go#L103)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Fixed
</blockquote></details>

Nit a switch here would be cleaner -- you could pull these error checks and the noop checks into it

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52090)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/alter_table_set_schema.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r1-132:-MDotlWJ6-8Unl9_GWks:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda966178eeb997e1b6fcf22633d2ccb2f33bb6a/pkg/sql/alter_table_set_schema.go#L132)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I'm thinking this setting the parent schema, and adding draining names at least
</blockquote></details>

Done.
___
*[pkg/sql/alter_table_set_schema.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDVvUjqEomyJE7fj1Zq:-MDotfww8BIaFLMy1gIF:b-j6ag6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/6495ac7c14d537d9e9b803122a559f6e095cd914/pkg/sql/alter_table_set_schema.go#L103)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Nit a switch here would be cleaner -- you could pull these error checks and the noop checks into it
</blockquote></details>

Made it into a switch, wasn't sure how to pull the noop check into the switch cases.

File: pkg/sql/alter_table_set_schema.go [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+		log.VEventf(ctx, 2, "CPut %s -> %d", newTbKey, tableDesc.ID)
+	}
+
+	if err := p.writeSchemaChange(
+		ctx, tableDesc, sqlbase.InvalidMutationID, tree.AsStringWithFQNames(n.n, params.Ann()),
+	); err != nil {
+		return err
+	}
+
+	b.CPut(newTbKey, tableDesc.ID, nil)
```
Comment (rohany) [inline pkg/sql/alter_table_set_schema.go:169]: move this cput and run closer to creation of the batch and key

File: pkg/sql/alter_table_set_schema.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+	}
+	tableDesc.AddDrainingName(renameDetails)
+
+	// Set the tableDesc's new schema id to the desired schema's id.
+	n.tableDesc.SetParentSchemaID(desiredSchemaID)
+
+	newTbKey := sqlbase.MakeObjectNameKey(ctx, p.ExecCfg().Settings,
+		databaseID, desiredSchemaID, tableDesc.Name).Key(p.ExecCfg().Codec)
+
+	// Update the descriptor to the schema. We leave the mapping of the old key
```
Comment (rohany) [inline pkg/sql/alter_table_set_schema.go:155]: I don't think this comment is necessary. It's also still confusing to me.

File: pkg/sql/parser/sql.y [hunk 12]
```diff
@@ -6176,6 +6185,48 @@ alter_rename_table_stmt:
     $$.val = &tree.RenameTable{Name: name, NewName: newName, IfExists: true, IsView: false}
   }
 
+alter_table_set_schema_stmt:
```
Comment (rohany) [inline pkg/sql/parser/sql.y:6188]: Make sure to add some parse tests in `TestParse`

**Review by rohany — APPROVED** [review]

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52090)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/alter_table_set_schema.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDVvUjqEomyJE7fj1Zq:-MDp2ctN7Ef0jfsdvfYm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6495ac7c14d537d9e9b803122a559f6e095cd914/pkg/sql/alter_table_set_schema.go#L103)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Made it into a switch, wasn't sure how to pull the noop check into the switch cases.
</blockquote></details>

Done.
___
*[pkg/sql/alter_table_set_schema.go, line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDVvverBtQ-kSL4m8_C:-MDp2djn42Ww7chWgprE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6495ac7c14d537d9e9b803122a559f6e095cd914/pkg/sql/alter_table_set_schema.go#L121)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

This is also used in rename_table as well, it can definitely be abstracted - I think I'll try to pull this out while adding the ALTER TYPE SET SCHEMA PR.
</blockquote></details>

Done.
___
*[pkg/sql/alter_table_set_schema.go, line 155 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r3-155:-MDp2ems15vmD4nBJfUu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b61902d45162b1f005e93169351b8fcb7c5a87f/pkg/sql/alter_table_set_schema.go#L155)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I don't think this comment is necessary. It's also still confusing to me.
</blockquote></details>

Done.
___
*[pkg/sql/alter_table_set_schema.go, line 169 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zs-r3-169:-MDp2goL6LPpx63s-NYR:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b61902d45162b1f005e93169351b8fcb7c5a87f/pkg/sql/alter_table_set_schema.go#L169)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

move this cput and run closer to creation of the batch and key
</blockquote></details>

Done
___
*[pkg/sql/parser/sql.y, line 6188 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52090#-MDRV1hqE-kD9Ddu29Zv-r3-6188:-MDp2j7h8pgUiR2jYi5u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b61902d45162b1f005e93169351b8fcb7c5a87f/pkg/sql/parser/sql.y#L6188)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Make sure to add some parse tests in `TestParse`
</blockquote></details>

Done.

**Comment by RichardJCai** [comment]
Thanks for the review!

**Comment by RichardJCai** [comment]
bors r=rohany

**Comment by RichardJCai** [comment]
Fixed merge conflict

**Comment by RichardJCai** [comment]
bors r=rohany

**Pull request merged**